[dtensor] add support for loss parallel

tianyu-l opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119877

Loss parallel is the last piece of sequence parallelism to enable. It enables efficient distributed cross entropy computation when the input is sharded on the class dimension (in a classification problem with many classes). The implementation is via a context manager `loss_parallel`, after enabling which users can directly use `torch.nn.functional.cross_entropy` or `torch.nn.CrossEntropyLoss` without modifying other parts of their code.

Here are the underlying rationales why we are going through these op replacements:

1. `nn.functional.cross_entropy` is the common method that OSS user is using for things like transformer training, to avoid changing user code, we want user to still use this function for loss calculation if they are already using it.
2. `nn.functional.cross_entropy` boils down into `aten.log_softmax` and `aten.nll_loss_foward/backward`, and DTensor now supports those ops already (#117723 #119255 #118917 #119256). They are doing computation with input *replicated* on the class dimension.
3. However when the input of this loss calculation is **sharded on the class dimension**, to run sharded computation efficiently, we need to run both `aten.log_softmax` and `aten.nll_loss_foward` with multiple all-reduce collectives **in the middle of** those aten ops. This is not possible if we are just overriding these two ops, so we need to have some way to **decompose** these two ops into smaller ops to have collectives run in the middle of these two ops.
4. We explored the existing decompositions (#118950). It seems working, except that `log_softmax_backward` and `nll_loss_backward` combined together in aten are implemented in a inefficient way, which would trigger an additional expensive collective. Recently some user also reported similar issues https://github.com/pytorch/pytorch/issues/119261.
5. Therefore, currently we are doing our own decomposition inside a context manager for sequence parallelism specifically. Once we have a better decomposition in core, we can possibly take that instead of reinventing the wheels here.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225

tianyu-l committed:
[dtensor] add support for loss parallel

[ghstack-poisoned]

wconstab reviewed (commented):
looks pretty good, and I assume this unblocks FSDP?

but I have a couple questions around the design

Why is a context manager needed (and why do we need to register special replacements for the operators in dispatch table) -- why can't we just add behaviors into DTensor's torch function so that when it handles these operators it calls these remap functions?

Is this more of a short term due to low confidence in the rewrite correctness, or is it actually not feasible to bake it into DTensor for some reason?

tianyu-l commented:
> Why is a context manager needed (and why do we need to register special replacements for the operators in dispatch table) -- why can't we just add behaviors into DTensor's torch function so that when it handles these operators it calls these remap functions?

Great questions! Here's just my understanding. Wanchao can say more on this.

For cross_entropy related ops, if loss parallel is not needed, we've implemented them in the normal way (#117723 #119255 #118917 #119256). When loss parallelism is needed, we need customized and somewhat slower op implementations in this PR. There has to be a way DTensor detects the intention of doing loss parallel or not. Using context manager gives the control to users; another way to do it involves automatically detecting if input is sharded on the dimension where loss parallel can be efficient -- I explored that path as well (in #118950) but that requires similar amount of customized code and/or even slower computation, in addition to moderate surgery to DTensor dispatching logic.

wanchaol commented:
@tianyu-l Please update the PR summary with more context and design on loss parallel.

@wconstab loss parallel is the last piece of sequence parallelism we want to enable. The underlying rationales why we are going through this op replacements instead of just support it via sharding propagation:
1. nn.functional.cross_loss_entropy is the common method that OSS user is using for things like transformer training, to avoid changing user code, we want user to still use this function for loss calculation if they are already using it.
2. `nn.functional.cross_loss_entropy` boils down into `aten.log_softmax` and `aten.nll_loss_foward/backward`, DTensor supports those ops now already.
3. However when the input of this loss calculation sharded on the vocab_dim, to run sharded computation efficiently, we need to run both `aten.log_softmax` and `aten.nll_loss_foward/backward` with multiple allreduce collectives **in the middle of" those aten ops. This is not possible if we are just overriding these two ops, we need to have some way to decompose these two ops to smaller ops to have collectives run in the middle of these two ops
4. We explored the decomposition way, and it seems working, except that because nll_loss_backward in aten is implemented in a in-efficient way, we would trigger additional collectives. Recently some user also reported similar issues https://github.com/pytorch/pytorch/issues/119261

Therefore we are doing this way of decomposition by our own currently for sequence parallelism specifically. Once we have a better decomposition in core, we can possibly take that instead of reinventing the wheels here

wanchaol commented on torch/distributed/tensor/parallel/loss_parallel.py:
```diff
@@ -0,0 +1,402 @@
+    ):
+        # The log_softmax backward computation has been performed during
+        # nll_loss backward in _nll_loss_and_log_softmax_backward.
+        # grad_input = grad_output - torch.exp(output) * torch.sum(
```
nit: this code comment is not needed?

wanchaol commented on torch/distributed/tensor/parallel/loss_parallel.py:
```diff
@@ -0,0 +1,402 @@
+        # )
+        grad_input = grad_output
+
+        if grad_output.dtype != input_dtype:
```
is this dtype conversion needed here?

wanchaol commented on torch/distributed/tensor/parallel/loss_parallel.py:
```diff
@@ -0,0 +1,402 @@
+        return grad_input
+
+    spec = grad_output._spec
+    res = _log_softmax_backward_no_computation(
```
IMO we don't need to extract the local tensor here, could just pass the grad_out through this function?

wanchaol commented on torch/distributed/tensor/parallel/loss_parallel.py:
```diff
@@ -0,0 +1,402 @@
+
+    # NOTE: The implementation follows torch._decomp.decomposition._log_softmax,
+    # with all_reduce manually inserted to perform distributed computation.
+    def _log_softmax(x, dim, half_to_float, mesh, mesh_dim):
```
seems inconsistent that this `_log_softmax` is inside the handler function while the later `_nll_loss_forward` is out of the handler. To make them consistent, let's put both out of the handler

wanchaol commented on torch/distributed/tensor/parallel/loss_parallel.py:
```diff
@@ -0,0 +1,402 @@
+
+def _enable_custom_loss_ops():
+    DTensor._op_dispatcher._custom_op_handlers.update(customized_loss_ops)
+    DTensor._op_dispatcher._allow_implicit_replication = True
```
why we turn on implicit replication?

wanchaol commented on torch/distributed/tensor/parallel/loss_parallel.py:
```diff
@@ -0,0 +1,402 @@
+
+# NOTE: The implementation follows torch._decomp.decomposition._nll_loss_forward,
+# with customized communication inserted to perform distributed computation.
+def _nll_loss_forward(
```
thanks for implementing these decomp functions! 

I think we should add a function that's similar to `nn.functional.cross_entropy_loss` in this file, and just call it `cross_entropy_loss`, which the exact same args/kwargs (maybe have optional mesh/mesh_dim args, and only run allreduce when these are present)

Then we should write a test to compare the `nn.functional.cross_entropy_loss` and the function we have, to assert numerics.

This function we wrote can also be public, meaning that if user want to use ours directly, they can do that

tianyu-l replied:
That's certainly doable. Basically local `cross_entropy` just needs to call `_log_softmax` and `_nll_loss_forward`, and for each of them, we can toggle all_reduce depending on whether `mesh` & `mesh_dim` are present.

However, am I right that this local `cross_entropy` won't have backward support, because the all_reduce's cannot be handled? So is the point of having this to provide user with more efficient forward-only computation when they want to use loss parallel with plain torch.Tensor input?

wanchaol reviewed (commented):
First pass on this and it looks great! Please see inlined comments

tianyu-l commented on torch/distributed/tensor/parallel/loss_parallel.py:
```diff
@@ -0,0 +1,402 @@
+
+
+@contextlib.contextmanager
+def loss_parallel():
```
Re 1: The semantic of `cross_entropy` says the `vocab_dim` should always be 1 if input dim >= 2 and 0 otherwise. Currently I'm throwing exception if input is not sharded on `vocab_dim`, in which case there is no need to enable this context manager.

Re 2: Return torch.Tensor sounds easy, can we just don't wrap the results to DTensor in the last backward handler? I'm not sure how to turn inputs from torch.Tensor without user doing anything extra, since at least one input has to be DTensor to route into DTensor dispatcher?

tianyu-l committed:
Update on "[dtensor] add support for loss parallel"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add support for loss parallel"


Loss parallel is the last piece of sequence parallelism to enable. It enables efficient distributed cross entropy computation when the input is sharded on the class dimension (in a classification problem with many classes). The implementation is via a context manager `loss_parallel`, after enabling which users can directly use `torch.nn.functional.cross_entropy` or `torch.nn.CrossEntropyLoss` without modifying other parts of their code.

Here are the underlying rationales why we are going through these op replacements:

1. `nn.functional.cross_entropy` is the common method that OSS user is using for things like transformer training, to avoid changing user code, we want user to still use this function for loss calculation if they are already using it.
2. `nn.functional.cross_entropy` boils down into `aten.log_softmax` and `aten.nll_loss_foward/backward`, and DTensor now supports those ops already (#117723 #119255 #118917 #119256). They are doing computation with input *replicated* on the class dimension.
3. However when the input of this loss calculation is **sharded on the class dimension**, to run sharded computation efficiently, we need to run both `aten.log_softmax` and `aten.nll_loss_foward` with multiple all-reduce collectives **in the middle of** those aten ops. This is not possible if we are just overriding these two ops, so we need to have some way to **decompose** these two ops into smaller ops to have collectives run in the middle of these two ops.
4. We explored the existing decompositions (#118950). It seems working, except that `log_softmax_backward` and `nll_loss_backward` combined together in aten are implemented in a inefficient way, which would trigger an additional expensive collective. Recently some user also reported similar issues https://github.com/pytorch/pytorch/issues/119261.
5. Therefore, currently we are doing our own decomposition inside a context manager for sequence parallelism specifically. Once we have a better decomposition in core, we can possibly take that instead of reinventing the wheels here.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add support for loss parallel"


Loss parallel is the last piece of sequence parallelism to enable. It enables efficient distributed cross entropy computation when the input is sharded on the class dimension (in a classification problem with many classes). The implementation is via a context manager `loss_parallel`, after enabling which users can directly use `torch.nn.functional.cross_entropy` or `torch.nn.CrossEntropyLoss` without modifying other parts of their code.

Here are the underlying rationales why we are going through these op replacements:

1. `nn.functional.cross_entropy` is the common method that OSS user is using for things like transformer training, to avoid changing user code, we want user to still use this function for loss calculation if they are already using it.
2. `nn.functional.cross_entropy` boils down into `aten.log_softmax` and `aten.nll_loss_foward/backward`, and DTensor now supports those ops already (#117723 #119255 #118917 #119256). They are doing computation with input *replicated* on the class dimension.
3. However when the input of this loss calculation is **sharded on the class dimension**, to run sharded computation efficiently, we need to run both `aten.log_softmax` and `aten.nll_loss_foward` with multiple all-reduce collectives **in the middle of** those aten ops. This is not possible if we are just overriding these two ops, so we need to have some way to **decompose** these two ops into smaller ops to have collectives run in the middle of these two ops.
4. We explored the existing decompositions (#118950). It seems working, except that `log_softmax_backward` and `nll_loss_backward` combined together in aten are implemented in a inefficient way, which would trigger an additional expensive collective. Recently some user also reported similar issues https://github.com/pytorch/pytorch/issues/119261.
5. Therefore, currently we are doing our own decomposition inside a context manager for sequence parallelism specifically. Once we have a better decomposition in core, we can possibly take that instead of reinventing the wheels here.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add support for loss parallel"


Loss parallel is the last piece of sequence parallelism to enable. It enables efficient distributed cross entropy computation when the input is sharded on the class dimension (in a classification problem with many classes). The implementation is via a context manager `loss_parallel`, after enabling which users can directly use `torch.nn.functional.cross_entropy` or `torch.nn.CrossEntropyLoss` without modifying other parts of their code.

Here are the underlying rationales why we are going through these op replacements:

1. `nn.functional.cross_entropy` is the common method that OSS user is using for things like transformer training, to avoid changing user code, we want user to still use this function for loss calculation if they are already using it.
2. `nn.functional.cross_entropy` boils down into `aten.log_softmax` and `aten.nll_loss_foward/backward`, and DTensor now supports those ops already (#117723 #119255 #118917 #119256). They are doing computation with input *replicated* on the class dimension.
3. However when the input of this loss calculation is **sharded on the class dimension**, to run sharded computation efficiently, we need to run both `aten.log_softmax` and `aten.nll_loss_foward` with multiple all-reduce collectives **in the middle of** those aten ops. This is not possible if we are just overriding these two ops, so we need to have some way to **decompose** these two ops into smaller ops to have collectives run in the middle of these two ops.
4. We explored the existing decompositions (#118950). It seems working, except that `log_softmax_backward` and `nll_loss_backward` combined together in aten are implemented in a inefficient way, which would trigger an additional expensive collective. Recently some user also reported similar issues https://github.com/pytorch/pytorch/issues/119261.
5. Therefore, currently we are doing our own decomposition inside a context manager for sequence parallelism specifically. Once we have a better decomposition in core, we can possibly take that instead of reinventing the wheels here.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add support for loss parallel"


Loss parallel is the last piece of sequence parallelism to enable. It enables efficient distributed cross entropy computation when the input is sharded on the class dimension (in a classification problem with many classes). The implementation is via a context manager `loss_parallel`, after enabling which users can directly use `torch.nn.functional.cross_entropy` or `torch.nn.CrossEntropyLoss` without modifying other parts of their code.

Here are the underlying rationales why we are going through these op replacements:

1. `nn.functional.cross_entropy` is the common method that OSS user is using for things like transformer training, to avoid changing user code, we want user to still use this function for loss calculation if they are already using it.
2. `nn.functional.cross_entropy` boils down into `aten.log_softmax` and `aten.nll_loss_foward/backward`, and DTensor now supports those ops already (#117723 #119255 #118917 #119256). They are doing computation with input *replicated* on the class dimension.
3. However when the input of this loss calculation is **sharded on the class dimension**, to run sharded computation efficiently, we need to run both `aten.log_softmax` and `aten.nll_loss_foward` with multiple all-reduce collectives **in the middle of** those aten ops. This is not possible if we are just overriding these two ops, so we need to have some way to **decompose** these two ops into smaller ops to have collectives run in the middle of these two ops.
4. We explored the existing decompositions (#118950). It seems working, except that `log_softmax_backward` and `nll_loss_backward` combined together in aten are implemented in a inefficient way, which would trigger an additional expensive collective. Recently some user also reported similar issues https://github.com/pytorch/pytorch/issues/119261.
5. Therefore, currently we are doing our own decomposition inside a context manager for sequence parallelism specifically. Once we have a better decomposition in core, we can possibly take that instead of reinventing the wheels here.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add support for loss parallel"


Loss parallel is the last piece of sequence parallelism to enable. It enables efficient distributed cross entropy computation when the input is sharded on the class dimension (in a classification problem with many classes). The implementation is via a context manager `loss_parallel`, after enabling which users can directly use `torch.nn.functional.cross_entropy` or `torch.nn.CrossEntropyLoss` without modifying other parts of their code.

Here are the underlying rationales why we are going through these op replacements:

1. `nn.functional.cross_entropy` is the common method that OSS user is using for things like transformer training, to avoid changing user code, we want user to still use this function for loss calculation if they are already using it.
2. `nn.functional.cross_entropy` boils down into `aten.log_softmax` and `aten.nll_loss_foward/backward`, and DTensor now supports those ops already (#117723 #119255 #118917 #119256). They are doing computation with input *replicated* on the class dimension.
3. However when the input of this loss calculation is **sharded on the class dimension**, to run sharded computation efficiently, we need to run both `aten.log_softmax` and `aten.nll_loss_foward` with multiple all-reduce collectives **in the middle of** those aten ops. This is not possible if we are just overriding these two ops, so we need to have some way to **decompose** these two ops into smaller ops to have collectives run in the middle of these two ops.
4. We explored the existing decompositions (#118950). It seems working, except that `log_softmax_backward` and `nll_loss_backward` combined together in aten are implemented in a inefficient way, which would trigger an additional expensive collective. Recently some user also reported similar issues https://github.com/pytorch/pytorch/issues/119261.
5. Therefore, currently we are doing our own decomposition inside a context manager for sequence parallelism specifically. Once we have a better decomposition in core, we can possibly take that instead of reinventing the wheels here.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add support for loss parallel"


Loss parallel is the last piece of sequence parallelism to enable. It enables efficient distributed cross entropy computation when the input is sharded on the class dimension (in a classification problem with many classes). The implementation is via a context manager `loss_parallel`, after enabling which users can directly use `torch.nn.functional.cross_entropy` or `torch.nn.CrossEntropyLoss` without modifying other parts of their code.

Here are the underlying rationales why we are going through these op replacements:

1. `nn.functional.cross_entropy` is the common method that OSS user is using for things like transformer training, to avoid changing user code, we want user to still use this function for loss calculation if they are already using it.
2. `nn.functional.cross_entropy` boils down into `aten.log_softmax` and `aten.nll_loss_foward/backward`, and DTensor now supports those ops already (#117723 #119255 #118917 #119256). They are doing computation with input *replicated* on the class dimension.
3. However when the input of this loss calculation is **sharded on the class dimension**, to run sharded computation efficiently, we need to run both `aten.log_softmax` and `aten.nll_loss_foward` with multiple all-reduce collectives **in the middle of** those aten ops. This is not possible if we are just overriding these two ops, so we need to have some way to **decompose** these two ops into smaller ops to have collectives run in the middle of these two ops.
4. We explored the existing decompositions (#118950). It seems working, except that `log_softmax_backward` and `nll_loss_backward` combined together in aten are implemented in a inefficient way, which would trigger an additional expensive collective. Recently some user also reported similar issues https://github.com/pytorch/pytorch/issues/119261.
5. Therefore, currently we are doing our own decomposition inside a context manager for sequence parallelism specifically. Once we have a better decomposition in core, we can possibly take that instead of reinventing the wheels here.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/parallel/loss.py:
```diff
@@ -0,0 +1,511 @@
+    grad_input = grad_output
+
+    if grad_output.dtype != input_dtype:
+        grad_input = grad_input.to(input_dtype)
```
is this dtype casting the only thing we need here? If so, I think we can get rid of this function and put it directly in the `_log_softmax_backward_handler`?

wanchaol commented on torch/distributed/tensor/parallel/loss.py:
```diff
@@ -0,0 +1,511 @@
+        >>> with loss_parallel():
+        >>>     loss = F.cross_entropy(dist_input, target, reduction="mean")
+        >>>     loss.backward()
+        >>> ...
```
we should probably add a warning at the end that tell user loss parallel API is experimental and subject to change:
```
.. warning ::
    The loss_parallel API is experimental and subject to change.
```

tianyu-l replied:
It seems we already have one at the top of Tensor Parallelism: "Tensor Parallelism APIs are experimental and subject to change." I feel adding this might make it a bit redundant.

wanchaol reviewed (approved):
Awesome work!! Have some extra comments inlined.

Didn't look into the math implementation details but assuming it's the same as our offline discussions and following the decomposition math logic already

tianyu-l committed:
Update on "[dtensor] add support for loss parallel"


Loss parallel is the last piece of sequence parallelism to enable. It enables efficient distributed cross entropy computation when the input is sharded on the class dimension (in a classification problem with many classes). The implementation is via a context manager `loss_parallel`, after enabling which users can directly use `torch.nn.functional.cross_entropy` or `torch.nn.CrossEntropyLoss` without modifying other parts of their code.

Here are the underlying rationales why we are going through these op replacements:

1. `nn.functional.cross_entropy` is the common method that OSS user is using for things like transformer training, to avoid changing user code, we want user to still use this function for loss calculation if they are already using it.
2. `nn.functional.cross_entropy` boils down into `aten.log_softmax` and `aten.nll_loss_foward/backward`, and DTensor now supports those ops already (#117723 #119255 #118917 #119256). They are doing computation with input *replicated* on the class dimension.
3. However when the input of this loss calculation is **sharded on the class dimension**, to run sharded computation efficiently, we need to run both `aten.log_softmax` and `aten.nll_loss_foward` with multiple all-reduce collectives **in the middle of** those aten ops. This is not possible if we are just overriding these two ops, so we need to have some way to **decompose** these two ops into smaller ops to have collectives run in the middle of these two ops.
4. We explored the existing decompositions (#118950). It seems working, except that `log_softmax_backward` and `nll_loss_backward` combined together in aten are implemented in a inefficient way, which would trigger an additional expensive collective. Recently some user also reported similar issues https://github.com/pytorch/pytorch/issues/119261.
5. Therefore, currently we are doing our own decomposition inside a context manager for sequence parallelism specifically. Once we have a better decomposition in core, we can possibly take that instead of reinventing the wheels here.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l commented:
@pytorchbot merge
